tests/src/test/scala/common/rest/SwaggerValidator.scala

```diff
@@ -0,0 +1,109 @@
+import akka.util.ByteString
+import com.atlassian.oai.validator.SwaggerRequestResponseValidator
+import com.atlassian.oai.validator.model.SimpleRequest
+import com.atlassian.oai.validator.model.SimpleResponse
+import com.atlassian.oai.validator.report.ValidationReport
+import com.atlassian.oai.validator.whitelist.ValidationErrorsWhitelist
+import com.atlassian.oai.validator.whitelist.rule.WhitelistRules
+
+trait SwaggerValidator {
+  private val specWhitelist = ValidationErrorsWhitelist
+    .create()
+    .withRule(
+      "Ignore action and trigger payloads",
+      WhitelistRules.allOf(
+        WhitelistRules.messageContains("Object instance has properties which are not allowed by the schema"),
+        WhitelistRules.anyOf(
+          WhitelistRules.pathContains("/web/"),
+          WhitelistRules.pathContains("/actions/"),
+          WhitelistRules.pathContains("/triggers/")),
+        WhitelistRules.methodIs(io.swagger.models.HttpMethod.POST)))
+    .withRule(
+      "Ignore invalid action kinds",
+      WhitelistRules.allOf(
+        WhitelistRules.messageContains("kind"),
+        WhitelistRules.messageContains("Instance value"),
+        WhitelistRules.messageContains("not found"),
+        WhitelistRules.pathContains("/actions/"),
+        WhitelistRules.methodIs(io.swagger.models.HttpMethod.PUT)))
+    .withRule(
+      "Ignore tests that check for invalid DELETEs and PUTs on actions",
+      WhitelistRules.anyOf(
+        WhitelistRules.messageContains("DELETE operation not allowed on path '/api/v1/namespaces/_/actions/'"),
+        WhitelistRules.messageContains("PUT operation not allowed on path '/api/v1/namespaces/_/actions/'")))
+
+  private val specValidator = SwaggerRequestResponseValidator
+    .createFor("apiv1swagger.json")
+    .withWhitelist(specWhitelist)
+    .build()
+
+  /**
+   * Validate a HTTP request and response against the Swagger spec. Request
+   * and response bodies are passed separately so that this validation
+   * does not have to consume the body content directly from the request
+   * and response, which would prevent callers from later consuming it.
+   *
+   * @param request the HttpRequest
+   * @param requestBody the request body
+   * @param response the HttpResponse
+   * @param responseBody the response body
+   * @return The list of validation error messages, if any
+   */
+  def validateRequestAndResponse(request: HttpRequest,
+                                 requestBody: Option[String],
+                                 response: HttpResponse,
+                                 responseBody: String): Seq[String] = {
+    var specRequestBuilder = new SimpleRequest.Builder(request.method.value, request.uri.path.toString())
+    for (header <- request.headers) {
+      specRequestBuilder = specRequestBuilder.withHeader(header.name, header.value)
+    }
+    for ((key, value) <- request.uri.query().toMap) {
+      specRequestBuilder = specRequestBuilder.withQueryParam(key, value)
+    }
```

Comment:
Scala tip: You can get around the variable here by using `foldLeft`, like:

```scala
val specRequest = {
  val builder = new SimpleRequest.Builder(request.method.value, request.uri.path.toString)
  val withHeaders = request.headers.foldLeft(builder)((b, header) => b.withHeader(header.name, header.value))
  val andQuery =
    request.uri.query().foldLeft(withHeaders) { case (b, (key, value)) => b.withQueryParam(key, value) }
  val andBody = requestBody
    .map(builder.withBody(_).withHeader("Content-Type", request.entity.contentType.value))
    .getOrElse(andQuery)

  andBody.build()
}
```

`foldLeft` runs over a list and applies each element to a seed value that you pass to it.

This also gets around using `Option.get`, which can throw an exception. Using combinators on Option (like map and getOrElse) you get a fully compiletime checked version of all the types without needing to wrap them in if/else statements.